nixos/modules/programs/wayland/niri.nix

```diff
@@ -0,0 +1,90 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.programs.niri;
+in
+{
+  options.programs.niri = {
+    enable = lib.mkEnableOption "Niri, a scrollable-tiling Wayland compositor";
+
+    package = lib.mkPackageOption pkgs "niri" { };
+
+    # Recommended by upstream
+    # https://github.com/YaLTeR/niri/wiki/Important-Software#authentication-agent
+    polkitAgent.enable = lib.mkEnableOption "a PolicyKit agent service" // {
+      default = config.security.polkit.enable;
+      defaultText = "config.security.polkit.enable";
+    };
+
+    # Recommended by upstream
+    # https://github.com/YaLTeR/niri/wiki/Important-Software#portals
+    portalPackage = lib.mkPackageOption pkgs "xdg-desktop-portal-gnome" {
+      # https://github.com/NixOS/nixpkgs/pull/348193#discussion_r1798515394
+      extraDescription = "Note: `xdg-desktop-portal-gnome` is required for screencast support";
+    };
```

Comment:
> it may still be nice to have at the user's choice for things like the file chooser

This also allows for easily using a modified version of `xdg-deskop-portal-gnome`; I don't see much harm in it with this note

Reply:
> it may still be nice to have at the user's choice for things like the file chooser

this should be done by adding the other portal in question, and setting `xdg.portal.config.niri."
org.freedesktop.impl.portal.FileChooser" = [ "lxqt" ];`.

changing the portal in this option will just break screencast functionality. you don't need to swap out the whole portal if you prefer one file chooser over another, they can coexist.

> This also allows for easily using a modified version of `xdg-deskop-portal-gnome`

fair enough, but given that [even the *gnome module* doesn't include this option](https://github.com/NixOS/nixpkgs/blob/4c2fcb090b1f3e5b47eaa7bd33913b574a11e0a0/nixos/modules/services/x11/desktop-managers/gnome.nix#L283), i don't see it being used by any niri user, ever. and if a user wants to change the gnome portal, they can always use an overlay to do so. (and if they're enabling the niri and gnome environments at the same time, they probably *should* use an overlay to do so, else they'll have two distinct `xdg-desktop-portal-gnome` installed, and that probably doesn't work out too well, if the system even builds at all)

Reply:
> changing the portal in this option will just break screencast functionality

Keep in mind that functionality can be disabled

> they can always use an overlay to do so

I don't think that should be the only option, for the same reasoning we have `programs.niri.package`

Reply:
I don't think it adds a lot of complexity, and it can indeed be useful for people to be able to modify the derivation.
Even if they'd have to stay with the gnome portal, someone might want to use a different version, or build it differently.

Reply:
> and i'd imagine if you're testing some custom version of xdp-gnome, it is probably common to want to test it with not just niri but also GNOME.
>
> in particular, enabling niri from this PR and changing the xdp-gnome version using this option alongside enabling GNOME causes the following build failure:

I'm sure there are countless other incompatibilities when enabling specific options together in nixpkgs. I don't think this should be a blocker over that, really

> as such, i struggle to see this option ever being useful for anyone?

It'd be nice for Niri users who want to modify the GNOME portal package easily without an overlay and not potentially cause mass rebuilds -- again, like `programs.niri.package`

> the ability to swap it out for a different portal implementation is bad

I mean that's just a side effect of has customizability. Are overlays also a bad solution because you could do `xdg-desktop-portal-gnome = final.xdg-desktop-portal-wlr`?

> it's rather just kinda confusing and makes it trivial to outright lose capabilities

That's why there's a note explaining what the GNOME implementation is required for

> currently basically just screencast

Which is an optional feature

> but in practice that is going to conflict with the GNOME module and cause your system build to fail

I don't know how many people are enabling both Niri and GNOME at once "in practice"

> so the only time i see this option making sense if if you're working on xdp-gnome or trying out someone else's work

If you're just trying out someone's work, it makes sense that you wouldn't test on GNOME. You also might not be working on the package at all, and just upgrading it to get a new feature early or pinning it to a specific version to avoid bugs

If you *really* want this removed though, I'll just do it at this point to get this merged